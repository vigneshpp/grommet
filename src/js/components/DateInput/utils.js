// Converting between Date and String types is handled via a "schema".
// The schema is an array of strings, split into strings with identical
// characters. So, 'mm/dd/yyyy' will be ['mm', '/', 'dd', '/', 'yyyyy'].

export const formatToSchema = format => {
  if (!format) return undefined;
  const result = [];

  let i = 0;
  let part;
  while (i < format.length) {
    if (!part || part[0] !== format[i]) {
      if (part) result.push(part);
      part = format[i];
    } else {
      part += format[i];
    }
    i += 1;
  }
  if (part) result.push(part);

  return result;
};

// convert value into text representation using the schema
export const valueToText = (value, schema) => {
  if (!value) return '';
  let text = '';

  const dates = (Array.isArray(value) ? value : [value]).map(v => new Date(v));

  let dateIndex = 0;
  let parts = {};
  schema.forEach(part => {
    const char = part[0].toLowerCase();
    // advance dateIndex if we already have this part
    while (
      dateIndex < dates.length &&
      (Number.isNaN(dates[dateIndex].date) ||
        ((char === 'm' || char === 'd' || char === 'y') && parts[char]))
    ) {
      dateIndex += 1;
      parts = {};
    }
    const date = dates[dateIndex];

    if (date && char === 'm') {
      text += date.getMonth() + 1;
      parts[char] = true;
    } else if (date && char === 'd') {
      text += date.getDate();
      parts[char] = true;
    } else if (date && char === 'y') {
      text += date.getFullYear();
      parts[char] = true;
    } else text += part;
  });

  return text;
};

const charCodeZero = '0'.charCodeAt(0);
const charCodeNine = '9'.charCodeAt(0);

const pullDigits = (text, index) => {
  let end = index;
  while (
    text.charCodeAt(end) >= charCodeZero &&
    text.charCodeAt(end) <= charCodeNine
  )
    end += 1;
  return text.slice(index, end);
};

export const textToValue = (text, schema) => {
  if (!text) return undefined;

  let result;
  const addDate = parts => {
    // do a little sanity checking on the values
    if (
      !parts.m ||
      !parts.d ||
      !parts.y ||
      parts.y.length < 4 ||
      parts.m.length > 2 ||
      parts.d.length > 2 ||
      parts.m > 12 ||
      parts.d > 31
    )
      return parts;
    const date = new Date(parts.y, parts.m - 1, parts.d).toISOString();
    if (!result) result = date;
    // single
    else if (Array.isArray(result)) result.push(date);
    // second
    else result = [result, date]; // third and beyond, unused?
    return {};
  };

  let parts = {};
  let index = 0;
  schema.forEach(part => {
    if (index < text.length) {
      const char = part[0].toLowerCase();
      if (parts[char] !== undefined) parts = addDate(parts);

      if (char === 'm') {
        parts.m = pullDigits(text, index);
        index += parts.m.length;
      } else if (char === 'd') {
        parts.d = pullDigits(text, index);
        index += parts.d.length;
      } else if (char === 'y') {
        parts.y = pullDigits(text, index);
        index += parts.y.length;
      } else if (text.slice(index, index + part.length) === part) {
        index += part.length;
      } else {
        // syntax error
        index = text.length;
        result = undefined;
      }
    }
  });
  addDate(parts);

  return result;
};
