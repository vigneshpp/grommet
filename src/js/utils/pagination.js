import { useCallback, useEffect, useState } from 'react';

export const normalizeShow = (showProp, step) => {
  let page;
  // by default, show refers to the index of an item,
  // but if using pagination, show can take the form of { page: # },
  // where page refers to the page # to show
  if (typeof showProp === 'number') page = Math.ceil((showProp + 1) / step);
  else if (typeof showProp === 'object' && 'page' in showProp)
    page = showProp.page;

  return page;
};

export const usePagination = ({ data, page, step, ...rest }) => {
  const totalPages = data ? Math.ceil(data.length / step) : 0;
  const [activePage, setActivePage] = useState(Math.min(page, totalPages) || 1);

  const itemsBeginIndex = step * (activePage - 1);
  const itemsEndIndex = itemsBeginIndex + step;
  const getCurrentItems = useCallback(
    items => {
      if (Array.isArray(items)) {
        return items.length
          ? items.slice(itemsBeginIndex, itemsEndIndex)
          : undefined;
      }
      return items;
    },
    [itemsBeginIndex, itemsEndIndex],
  );
  const [currentItems, setCurrentItems] = useState(getCurrentItems(data));

  useEffect(() => {
    if (data) {
      setCurrentItems(getCurrentItems(data));
    }
  }, [data, getCurrentItems, setCurrentItems]);

  const paginationProps = {
    numberItems: data && data.length,
    onChange: event => setActivePage(event.page),
    page,
    step,
    ...rest, // let anything coming from paginate prop override component
  };

  return [currentItems, paginationProps];
};
