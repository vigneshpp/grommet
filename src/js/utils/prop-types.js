import { PropTypes } from 'react-desc';

export const a11yTitlePropType = PropTypes.string.description(
  `Custom label to be used by screen readers. When provided, an aria-label will
   be added to the element.`,
);

export const colorPropType = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({ dark: PropTypes.string, light: PropTypes.string }),
]);

export const backgroundDoc = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.shape({
    color: colorPropType,
    dark: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    image: PropTypes.string,
    position: PropTypes.string,
    opacity: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.oneOf(['weak', 'medium', 'strong']),
    ]),
    repeat: PropTypes.oneOfType([
      PropTypes.oneOf(['no-repeat', 'repeat']),
      PropTypes.string,
    ]),
    size: PropTypes.oneOfType([
      PropTypes.oneOf(['cover', 'contain']),
      PropTypes.string,
    ]),
    light: PropTypes.string,
  }),
]).description(`Either a color 
identifier to use for the background color. For example: 'neutral-1'. Or, a 
'url()' for an image. Dark is not needed if color is provided.`);

export const MARGIN_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
];

export const marginProp = PropTypes.oneOfType([
  PropTypes.oneOf(['none', ...MARGIN_SIZES]),
  PropTypes.shape({
    bottom: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    end: PropTypes.oneOfType([PropTypes.oneOf(MARGIN_SIZES), PropTypes.string]),
    horizontal: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    left: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    right: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    start: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
    top: PropTypes.oneOfType([PropTypes.oneOf(MARGIN_SIZES), PropTypes.string]),
    vertical: PropTypes.oneOfType([
      PropTypes.oneOf(MARGIN_SIZES),
      PropTypes.string,
    ]),
  }),
  PropTypes.string,
]).description(`The amount of margin around the component. An object can
    be specified to distinguish horizontal margin, vertical margin, and
    margin on a particular side.`);

const PAD_SIZES = ['xxsmall', 'xsmall', 'small', 'medium', 'large', 'xlarge'];

export const padPropType = PropTypes.oneOfType([
  PropTypes.oneOf(['none', ...PAD_SIZES]),
  PropTypes.shape({
    bottom: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    end: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    horizontal: PropTypes.oneOfType([
      PropTypes.oneOf(PAD_SIZES),
      PropTypes.string,
    ]),
    left: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    right: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    start: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    top: PropTypes.oneOfType([PropTypes.oneOf(PAD_SIZES), PropTypes.string]),
    vertical: PropTypes.oneOfType([
      PropTypes.oneOf(PAD_SIZES),
      PropTypes.string,
    ]),
  }),
  PropTypes.string,
])
  .description(
    `The amount of padding around the box contents. An
    object can be specified to distinguish horizontal padding, vertical
    padding, and padding on a particular side of the box`,
  )
  .defaultValue('none');

export const genericProps = {
  a11yTitle: a11yTitlePropType,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch'])
    .description(`How to align along the cross axis when contained in
      a Box or along the column axis when contained in a Grid.`),
  gridArea: PropTypes.string.description(`The name of the area to place
    this inside a parent Grid.`),
  margin: marginProp,
};

export const hoverIndicatorPropType = PropTypes.oneOfType([
  PropTypes.bool,
  PropTypes.string,
  PropTypes.oneOf(['background']),
  PropTypes.shape({
    color: PropTypes.string,
    dark: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    image: PropTypes.string,
    light: PropTypes.string,
    position: PropTypes.string,
    opacity: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool,
      PropTypes.number,
      PropTypes.oneOf(['weak', 'medium', 'strong']),
    ]),
    repeat: PropTypes.oneOfType([
      PropTypes.oneOf(['no-repeat', 'repeat']),
      PropTypes.string,
    ]),
    size: PropTypes.oneOfType([
      PropTypes.oneOf(['cover', 'contain']),
      PropTypes.string,
    ]),
  }),
]);

export const pointPropType = PropTypes.oneOf([
  'circle',
  'diamond',
  'square',
  'star',
  'triangle',
  'triangleDown',
]);
