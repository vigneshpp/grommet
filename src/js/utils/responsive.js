export const getBreakpoint = (windowWidth, theme) => {
  let result;
  const sortedBreakpoints = Object.keys(theme.global.breakpoints).sort(
    (a, b) => {
      const first = theme.global.breakpoints[a];
      const second = theme.global.breakpoints[b];
      if (!first) return 1;
      if (!second) return -1;
      if (!first.value) return 1;
      if (!second.value) return -1;
      return first.value - second.value;
    },
  );
  // the last breakpoint on the sorted array should have
  // no windowWidth boundaries
  const lastBreakpoint = sortedBreakpoints[sortedBreakpoints.length - 1];
  sortedBreakpoints.some(name => {
    const breakpoint = theme.global.breakpoints[name];
    if (breakpoint) {
      if (!breakpoint.value || breakpoint.value >= windowWidth) {
        result = name;
        return true;
      }
    }
    return false;
  });
  return result || lastBreakpoint;
};

export const getDeviceBreakpoint = (type, theme) =>
  theme.global.deviceBreakpoints[type];
