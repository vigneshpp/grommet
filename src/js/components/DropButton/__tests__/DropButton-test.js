import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';
import 'jest-axe/extend-expect';
import 'regenerator-runtime/runtime';

import { createPortal, expectPortal } from '../../../utils/portal';

import { DropButton } from '..';

describe('DropButton', () => {
  beforeEach(createPortal);

  afterEach(cleanup);

  test('should have no accessibility violations', async () => {
    const { container } = render(
      <DropButton
        a11yTitle="test"
        dropContent={<div id="drop-contents">drop contents</div>}
      />,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('closed', () => {
    const component = renderer.create(
      <DropButton
        label="Dropper"
        dropContent={<div id="drop-contents">drop contents</div>}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('opened', () => {
    const component = renderer.create(
      <DropButton
        label="Dropper"
        open
        dropContent={<div id="drop-contents">drop contents</div>}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('open and close', () => {
    window.scrollTo = jest.fn();
    const { getByText, container } = render(
      <DropButton
        label="Dropper"
        dropContent={<div id="drop-contents">Drop Contents</div>}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(document.getElementById('drop-contents')).toBeNull();

    fireEvent.click(getByText('Dropper'));
    expectPortal('drop-contents').toMatchSnapshot();

    fireEvent.click(getByText('Dropper'));
    expect(document.getElementById('drop-contents')).toBeNull();
    expect(window.scrollTo).toBeCalled();
  });

  test('close by clicking outside', done => {
    const { getByText, container } = render(
      <DropButton
        label="Dropper"
        dropContent={<div id="drop-contents">Drop Contents</div>}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(document.getElementById('drop-contents')).toBeNull();

    fireEvent.click(getByText('Dropper'));
    expectPortal('drop-contents').toMatchSnapshot();

    fireEvent(
      document,
      new MouseEvent('mousedown', { bubbles: true, cancelable: true }),
    );

    setTimeout(() => {
      expect(document.getElementById('drop-contents')).toBeNull();
      done();
    }, 50);
  });

  test('disabled', () => {
    const { getByText, container } = render(
      <DropButton
        disabled
        label="Dropper"
        dropContent={<div id="drop-contents">Drop Contents</div>}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(document.getElementById('drop-contents')).toBeNull();

    fireEvent.click(getByText('Dropper'));
    expect(document.getElementById('drop-contents')).toBeNull();
  });

  test('opened ref', () => {
    const ref = React.createRef();
    const { container } = render(
      <DropButton
        ref={ref}
        open
        label="Dropper"
        dropContent={<div id="drop-contents">Drop Contents</div>}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    expectPortal('drop-contents').toMatchSnapshot();
  });

  test('ref function', () => {
    const ref = jest.fn();
    const { container } = render(
      <DropButton
        ref={ref}
        open
        label="Dropper"
        dropContent={<div id="drop-contents">Drop Contents</div>}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(ref).toBeCalled();
    expectPortal('drop-contents').toMatchSnapshot();
  });
});
