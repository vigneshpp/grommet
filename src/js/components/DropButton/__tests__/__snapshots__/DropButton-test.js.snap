// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropButton close by clicking outside 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

<button
  aria-label="Open Drop"
  class="c0"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton close by clicking outside 2`] = `
@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {

}

<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton close by clicking outside 3`] = `""`;

exports[`DropButton closed 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

<button
  aria-label="Open Drop"
  className="c0"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton disabled 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  opacity: 0.3;
  cursor: default;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

<button
  aria-label="Open Drop"
  class="c0"
  disabled=""
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton open and close 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

<button
  aria-label="Open Drop"
  class="c0"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton open and close 2`] = `
@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {

}

<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton open and close 3`] = `""`;

exports[`DropButton opened 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

<button
  aria-label="Open Drop"
  className="c0"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton opened ref 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {

}

<button
  aria-label="Open Drop"
  class="c0"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton opened ref 2`] = `
@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {

}

<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton opened ref 3`] = `""`;

exports[`DropButton ref function 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {

}

<button
  aria-label="Open Drop"
  class="c0"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton ref function 2`] = `
@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {

}

<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton ref function 3`] = `""`;

exports[`DropButton should have no accessibility violations 1`] = `
.c0 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus > circle,
.c0:focus > ellipse,
.c0:focus > line,
.c0:focus > path,
.c0:focus > polygon,
.c0:focus > polyline,
.c0:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c0:focus::-moz-focus-inner {
  border: 0;
}

<button
  aria-label="test"
  class="c0"
  type="button"
/>
`;
